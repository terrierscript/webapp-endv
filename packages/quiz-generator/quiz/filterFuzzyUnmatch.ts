import nlp from "compromise"
import { isTruthy } from "typesafe-utils"


export const filterFuzzyUnmatch = (word: string, target: string[]) => {
  const n = nlp(word)
  const collects = target.filter(l => {
    const fuzzyMatch = l.split(" ").some(ll => {
      // const match = n.match(`~${ll}~`, null, { fuzzy: 0.3 })
      const match = n.match(ll, { fuzzy: 0.3 })
      // console.log({ match, word, l, ll })
      // @ts-ignore
      return match.length > 0
    })
    return !fuzzyMatch
  })
  return collects
}

export function* filterFuzzyUnmatchGenerator(word: string, target: string[]) {
  const n = nlp(word)

  for (let i = 0;i < target.length;i++) {
    const l = target[i]
    // console.log(i, target[i])
    const fuzzyMatch = l.split(" ").some(ll => {
      const match = n.match(ll, { fuzzy: 0.3 })

      // console.log({ match, word, l, ll })
      // @ts-ignore
      return match.length > 0
    })
    if (!fuzzyMatch) {
      yield l
    }
  }
}

const g = filterFuzzyUnmatchGenerator("a", ["cc"])


export const filterFuzzyUnmatchNum = (word: string, target: string[], num: number): string[] => {
  const generator = filterFuzzyUnmatchGenerator(word, target)
  return [...Array(num)]
    .map((): string | null => {
      const r = generator.next()
      if (typeof r.value === "string") {
        return r.value
      }
      return null
    })
    .filter(isTruthy)
}
