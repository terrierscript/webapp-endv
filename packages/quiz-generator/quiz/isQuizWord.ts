export const isQuizWord = (f: string) => {
  if (!f)
    return false
  // if (f === word) return false
  // if (f.split(" ").length !== 1) return false
  if (f.split("-").length !== 1)
    return false
  if (/[A-Z]/.test(f))
    return false
  if (/[0-9]/.test(f))
    return false
  return true
}
