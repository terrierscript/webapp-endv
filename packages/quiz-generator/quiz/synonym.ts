import { isTruthy } from "typesafe-utils"
import { isQuizWord } from "./isQuizWord"
import { getNestedLemma, NestedLemmaData } from "../nested/lemma"

export const getWordSynonyms = (word: string, rel: NestedLemmaData) => {
  const _currentLemmas: string[] = rel.lexicalEntry?.map(l => l.senses.map(se => {
    return se?.synset?.lemmas
  })).flat(2)
    .filter(isTruthy) ?? []
  const currentLemmas = [...new Set(_currentLemmas)]
    .filter(isQuizWord)
  return currentLemmas
}

export const getSynonym = (word: string) => {
  const lemma = getNestedLemma(word)
  const synonym = getWordSynonyms(word, lemma)
  return synonym.filter(x => x !== word)
}
