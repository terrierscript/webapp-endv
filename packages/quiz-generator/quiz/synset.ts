import { isTruthy } from "typesafe-utils"
import { isQuizWord } from "./isQuizWord"
import { NestedLemmaData } from "../nested/lemma"
import { getNestedSynset } from "../nested/synset"
import { RelationRecord } from "../nested/types"

const unique = (arr: (string | undefined)[]): string[] => {
  const filterd = arr.filter(isTruthy)
  return [
    ...new Set(filterd)
  ]
}

const filterParentAndChildren = (rels: RelationRecord[]) => {
  let parents: string[] = []
  let children: string[] = []
  rels.map(rel => {
    if (rel.relType === "hypernym") {
      parents = [...parents, ...rel.targets]
    }
    if (rel.relType === "hyponym") {
      children = [...children, ...rel.targets]
    }
  })
  return {
    parents: [...new Set(parents)].filter(isQuizWord),
    children: [...new Set(children)].filter(isQuizWord)
  }
}

export const digSynset = (rel: NestedLemmaData) => {
  const rels = rel?.lexicalEntry?.map(l => l.senses.map(se => {
    return se?.synset?.relations
  })).flat(2).filter(rel => !!rel)
    // @ts-ignore
    .map((rel: RelationRecord) => {
      return {
        ...rel,
        targets: unique(rel.targets.map(t => {
          return getNestedSynset(t)?.lemmas
        }).flat())
      }
    }) ?? []
  const pc = filterParentAndChildren(rels)
  return pc
}
