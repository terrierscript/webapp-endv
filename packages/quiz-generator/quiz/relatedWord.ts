import { getNestedLemma } from "../nested/lemma"
import { getSynonym } from "./synonym"
import { digSynset } from "./synset"
const levelIntersect = (targets: string[][]) => {
  const dups = new Set()
  const results: string[][] = []
  targets.forEach(t => {
    const f = t.filter(item => !dups.has(item))
    results.push(f)
    t.forEach(tt => dups.add(tt))
  })
  return results
}

export const getRelatedWords = (word: string) => {
  const lemma = getNestedLemma(word)
  const synonym = getSynonym(word)
  const { parents, children } = digSynset(lemma)
  return {
    word, synonym, parents, children
  }
}

export const relatedWord = (word: string) => {
  const { synonym, parents } = getRelatedWords(word)
  // const d1 = relatedWords(parents)
  // const d2 = relatedWords(d1.parents)
  const levels = [
    [word], synonym, parents,
    // d1.children, d2.children
  ]
  return levelIntersect(levels)
}
