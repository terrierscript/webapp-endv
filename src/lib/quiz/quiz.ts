import { RelationRecord } from "../dictionary/types"
import { getNestedLemma, NestedLemmaData } from "../nested/lemma"
import { getNestedSynset } from "../nested/synset"
import { isTruthy } from 'typesafe-utils'
import nlp from "compromise"

const isQuizWord = (f: string) => {
  if (!f)
    return false
  // if (f === word) return false
  if (f.split(" ").length !== 1)
    return false
  if (f.split("-").length !== 1)
    return false
  if (/[A-Z]/.test(f))
    return false
  return true

}
const getWordSynonyms = (word: string, rel: NestedLemmaData) => {
  const _currentLemmas: string[] = rel.lexicalEntry?.map(l => l.senses.map(se => {
    return se?.synset?.lemmas
  })).flat(2)
    .filter(isTruthy) ?? []
  const currentLemmas = [...new Set(_currentLemmas)]
    .filter(isQuizWord)
  return currentLemmas
}
const unique = (arr: (string | undefined)[]): string[] => [...new Set(arr.filter(isTruthy))]
const filterParentAndChildren = (rels: RelationRecord[]) => {
  let parents: string[] = []
  let children: string[] = []
  rels.map(rel => {
    if (rel.relType === "hypernym") {
      parents = [...parents, ...rel.targets]
    }
    if (rel.relType === "hyponym") {
      children = [...children, ...rel.targets]
    }
  })
  return {
    parents: [...new Set(parents)].filter(isQuizWord),
    children: [...new Set(children)].filter(isQuizWord)
  }
}
const digSynset = (rel: NestedLemmaData) => {
  const rels = rel.lexicalEntry?.map(l => l.senses.map(se => {
    return se?.synset?.relations
  })).flat(2).filter(rel => !!rel)
    // @ts-ignore
    .map((rel: RelationRecord) => {
      return {
        ...rel,
        targets: unique(rel.targets.map(t => {
          return getNestedSynset(t)?.lemmas
        }).flat())
      }
    }) ?? []
  const pc = filterParentAndChildren(rels)
  return pc
}
const intersect = (targetArr: string[], filterSet: Set<string>) => {
  return targetArr.filter(item => !filterSet.has(item))
}
const levelIntersect = (targets: string[][]) => {
  const dups = new Set()
  const results: string[][] = []
  targets.forEach(t => {
    const f = t.filter(item => !dups.has(item))
    results.push(f)
    t.forEach(tt => dups.add(tt))
  })
  return results
}
const getRelatedWords = (word: string) => {
  const lemma = getNestedLemma(word)
  const synonym = getWordSynonyms(word, lemma)
  const { parents, children } = digSynset(lemma)
  return {
    word, synonym, parents, children
  }
}
const addSet = (set: Set<string>, append: string[]) => {
  append.map(a => set.add(a))
  return set
}
const kinderWords = (words: string[]) => {
  const m: {
    [k in string]: Set<string>
  } = {
    word: new Set(),
    synonym: new Set(),
    parents: new Set(),
    children: new Set(),
  }
  words.map(w => {
    return getRelatedWords(w)
  }).forEach(({ word, synonym, parents, children }) => {
    m.word = addSet(m.word, [word])
    m.synonym = addSet(m.synonym, synonym)
    m.parents = addSet(m.parents, parents)
    m.children = addSet(m.children, children)
  })
  return Object.fromEntries(Object.entries(m).map(([k, v]) => {
    return [k, [...v]]
  }))
}
function random(arr: string[]) {
  return arr[Math.floor(Math.random() * arr.length)]
}
const relatedWord = (word: string) => {
  const { synonym, parents } = getRelatedWords(word)
  // const d1 = relatedWords(parents)
  // const d2 = relatedWords(d1.parents)
  const levels = [
    [word], synonym, parents,
    // d1.children, d2.children
  ]
  return levelIntersect(levels)
}
export type QuizSet = {
  word: string,
  collect: string,
  incollect: string,
}
const generateQuiz = (word: string): QuizSet | null => {
  const [w, l1, l2] = relatedWord(word.toString())
  const n = nlp(word)
  const collects = l1.filter(l => {
    // @ts-ignore
    const fuzzyMatch: number = n.match(l, { fuzzy: 0.5 }).length
    return fuzzyMatch == 0
  })
  const incollects = intersect(l2, new Set([...l1, ...w]))
  if (collects.length === 0 || incollects.length === 0) {
    return null
  }
  return {
    word: w[0],
    collect: random(collects),
    incollect: random(incollects)
  }
}

const shuffle = (arr: string[]) => arr.sort(() => Math.random() - .5)

export const generateQuizzes = (seedWord: string) => {
  const { parents } = getRelatedWords(seedWord)
  const kinder = kinderWords(parents)

  const cousinWords = shuffle([...new Set([...kinder.children])])
  const words = cousinWords.slice(0, 10)

  return words.map(w => generateQuiz(w)).filter(isTruthy).slice(0, 10)
}

