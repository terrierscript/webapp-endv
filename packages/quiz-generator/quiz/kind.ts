import { getRelatedWords } from "./relatedWord"

const addSet = (set: Set<string>, append: string[]) => {
  append.map(a => set.add(a))
  return set
}

export const kinderWords = (words: string[]) => {
  const m: {
    [k in string]: Set<string>
  } = {
    word: new Set(),
    synonym: new Set(),
    parents: new Set(),
    children: new Set(),
  }
  words.map(w => {
    return getRelatedWords(w)
  }).forEach(({ word, synonym, parents, children }) => {
    m.word = addSet(m.word, [word])
    m.synonym = addSet(m.synonym, synonym)
    m.parents = addSet(m.parents, parents)
    m.children = addSet(m.children, children)
  })
  return Object.fromEntries(Object.entries(m).map(([k, v]) => {
    return [k, [...v]]
  }))
}
