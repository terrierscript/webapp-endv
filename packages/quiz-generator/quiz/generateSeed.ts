import { relatedWord } from "./relatedWord"
import { kinderWords } from "./kind"
import { arrayIntersect } from "../util/arrayIntersect"
import { shuffle } from "../util/shuffle"
import { filterFuzzyUnmatch } from "./filterFuzzyUnmatch"
import { isQuizWord } from "./isQuizWord"
import dictionary from "../dictionary/dictionary"

type QuizSeed = {
  word: string,
  collects: string[],
  incollects: string[]
}

export const generateQuizSeed = (word: string): QuizSeed => {
  const [w, level1, level2] = relatedWord(word.toString())
  const d1 = kinderWords(level2)
  // shuffle(level2).slice(0, 5))
  const collects = level1
  const incollectCandidate = new Set([...level1, ...level2, ...w])
  const incollects = arrayIntersect(d1.children, incollectCandidate)
  return {
    word,
    collects,
    incollects,
    // debug: { l1, l2 }
  }
}

export const seedToQuiz = (seed: QuizSeed) => {
  const r = filterFuzzyUnmatch(seed.word, seed.collects)
  // console.log(seed)
  // if (r.le1) {
  //   return null
  // }
  const quiz = {
    word: seed.word,
    collect: shuffle(r)[0],
    incollects: shuffle(seed.incollects).slice(0, 3)
  }

  if (!quiz.collect || quiz.incollects.length < 3) {
    return null
  }
  return quiz
}

const generateRandom = () => {
  const word = dictionary.getRandomWord()
  if (!isQuizWord(word)) {
    return
  }
  if (word.split(" ").length > 1) {
    return
  }
  return word
}

export const generateRandomQuiz = () => {
  for (let i = 0;i < 100;i++) {
    const word = generateRandom()
    if (!word) {
      continue
    }
    const seed = generateQuizSeed(word)
    const quiz = seedToQuiz(seed)
    if (quiz) {
      return quiz
    }
  }
}