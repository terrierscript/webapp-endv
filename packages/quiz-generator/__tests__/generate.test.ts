import { expect, test } from "vitest"
import { generateQuizSeed } from "../quiz/generateSeed"

test("get synonym", () => {
  const data = generateQuizSeed("dog")
  expect(data).toMatchSnapshot()
})
test("get synonym", () => {
  const data = generateQuizSeed("cat")
  expect(data).toMatchSnapshot()
})