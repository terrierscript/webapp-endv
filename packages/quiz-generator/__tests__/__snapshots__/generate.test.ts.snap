// Vitest Snapshot v1

exports[`get synonym 1`] = `
{
  "collects": [
    "domestic dog",
    "frump",
    "heel",
    "cad",
    "bounder",
    "hound",
    "blackguard",
    "hotdog",
    "hot dog",
    "wiener",
    "wienerwurst",
    "frankfurter",
    "weenie",
    "frank",
    "pawl",
    "click",
    "detent",
    "firedog",
    "andiron",
    "chase",
    "trail",
    "go after",
    "tag",
    "tail",
    "chase after",
    "track",
    "give chase",
  ],
  "incollects": [
    "fang",
    "bitch",
    "wolf",
    "jackal",
    "wild dog",
    "hyena",
    "hyaena",
    "fox",
    "feeder",
    "stocker",
    "head",
    "house cat",
    "domestic cat",
    "stray",
    "bag",
    "old bag",
    "cow",
    "tartar",
    "dragon",
    "termagant",
    "shrew",
    "harpy",
    "hellcat",
    "vixen",
    "nudnick",
    "nudnik",
    "scorer",
    "escort",
    "date",
    "playfellow",
    "playmate",
    "tovarisch",
    "tovarich",
    "gallows bird",
    "knave",
    "varlet",
    "scallywag",
    "rapscallion",
    "scalawag",
    "rogue",
    "rascal",
    "villainess",
    "supervillain",
    "black pudding",
    "blood sausage",
    "blood pudding",
    "bologna",
    "chipolata",
    "chorizo",
    "headcheese",
    "knackwurst",
    "knockwurst",
    "liver sausage",
    "liver pudding",
    "liverwurst",
    "pepperoni",
    "pork sausage",
    "salami",
    "souse",
    "bench hook",
    "doorstop",
    "doorstopper",
    "trip",
    "tripper",
    "hasp",
    "hood latch",
    "hook",
    "latch",
    "fair catch",
    "interception",
    "reception",
    "rebound",
    "shoestring catch",
    "interlock",
    "mesh",
    "meshing",
    "interlocking",
    "harpoon",
    "fish",
    "nett",
    "net",
    "intercept",
    "retake",
    "recapture",
    "rope",
    "lasso",
    "snag",
    "run",
    "hunt",
    "hunt down",
    "track down",
    "frog",
    "batfowl",
    "rat",
    "trammel",
    "entrap",
    "snare",
    "trap",
    "ensnare",
    "get",
    "visualize",
    "visualise",
    "spectate",
    "preview",
    "binge",
    "rewatch",
    "catch cold",
    "hold",
    "work",
    "surcease",
    "cessation",
    "stand",
    "standstill",
    "standdown",
    "hemostasia",
    "hemostasis",
    "haemostasia",
    "haemostasis",
    "pit stop",
    "countercheck",
    "logjam",
    "bus stop",
    "checkpoint",
    "loading area",
    "loading zone",
    "way station",
    "stopover",
    "labial stop",
    "glottal plosive",
    "glottal catch",
    "glottal stop",
    "suction stop",
    "suspension point",
    "iris diaphragm",
    "iris",
    "breechblock",
    "breech closer",
    "stopple",
    "plug",
    "stopper",
    "vapor lock",
    "vapour lock",
    "go off",
    "pull up short",
    "check",
    "rein",
    "rein in",
    "conk",
    "stall",
    "pull up",
    "draw up",
    "haul up",
    "brake",
    "settle",
    "drop",
    "knock off",
    "leave off",
    "sign off",
    "withdraw",
    "retire",
    "pull the plug",
    "close off",
    "shut off",
    "cheese",
    "call it quits",
    "call it a day",
    "break",
    "unfollow",
    "embargo",
    "stay",
    "call",
    "lay over",
    "stop over",
    "arrest",
    "halt",
    "bring up",
    "cut",
    "flag down",
    "fracture",
    "bog down",
    "bog",
    "cut off",
    "break up",
    "interrupt",
    "disrupt",
    "cut out",
    "cut down",
    "pass away",
    "lapse",
    "go out",
    "recess",
    "adjourn",
    "vanish",
    "disappear",
    "culminate",
    "climax",
    "run out",
    "go",
    "run low",
    "run short",
    "go away",
    "close",
    "conclude",
    "turn out",
    "come out",
    "discontinue",
    "block off",
    "sustainment",
    "sustenance",
    "upkeep",
    "maintenance",
    "sustentation",
    "logistic assistance",
    "logistic support",
    "attachment",
    "adhesion",
    "adherence",
    "protagonism",
    "advocacy",
    "sponsorship",
    "indorsement",
    "endorsement",
    "approval",
    "blessing",
    "approving",
    "reassurance",
    "mainstay",
    "keystone",
    "linchpin",
    "backbone",
    "anchor",
    "lynchpin",
    "lifeline",
    "close support",
    "amenities",
    "comforts",
    "conveniences",
    "creature comforts",
    "meal ticket",
    "subsistence",
    "abutment",
    "architrave",
    "basement",
    "buttress",
    "buttressing",
    "base",
    "fundament",
    "foundation",
    "substructure",
    "understructure",
    "foot",
    "groundwork",
    "footstall",
    "plinth",
    "pedestal",
    "prop",
    "furket",
    "musket rest",
    "shoring",
    "propping up",
    "shoring up",
    "hanging",
    "suspension",
    "dangling",
    "discant",
    "descant",
    "vamp",
    "arch support",
    "backrest",
    "back",
    "backboard",
    "baluster",
    "bar",
    "bearing",
    "bearing wall",
    "bedpost",
    "bookend",
    "brace",
    "bracket",
    "wall bracket",
    "bridge",
    "footing",
    "foothold",
    "handrest",
    "hanger",
    "harness",
    "harp",
    "headstock",
    "leg",
    "perch",
    "pier",
    "pillow block",
    "rack",
    "rest",
    "rib",
    "rocker",
    "seat",
    "shelf",
    "radius",
    "spoke",
    "wheel spoke",
    "stair",
    "step",
    "stirrup iron",
    "stirrup",
    "gunstock",
    "stock",
    "structural member",
    "tailstock",
    "tee",
    "football tee",
    "undercarriage",
    "yoke",
    "sponsor",
    "frequent",
    "shop",
    "patronize",
    "patronise",
    "shop at",
    "buy at",
    "help",
    "aid",
    "assist",
    "boost",
    "advance",
    "promote",
    "further",
    "encourage",
    "root",
    "undergird",
    "indorse",
    "endorse",
    "second",
    "fund",
    "subsidize",
    "subsidise",
    "provide",
    "bring home the bacon",
    "see through",
    "warrant",
    "guarantee",
    "champion",
    "defend",
    "scaffold",
    "block",
    "carry",
    "chock",
    "buoy",
    "buoy up",
    "pole",
    "underpin",
    "prop up",
    "shore",
    "shore up",
    "truss",
    "back up",
    "vouch",
    "verify",
    "establish",
    "prove",
    "show",
    "demonstrate",
    "shew",
    "document",
    "validate",
    "rationalize",
    "rationalise",
    "apologize",
    "apologise",
    "excuse",
    "take someone's side",
    "take sides with",
    "stand up",
    "stick up",
    "uphold",
    "swallow",
    "accept",
    "live with",
    "stand for",
    "hold still for",
    "bear up",
    "take lying down",
    "take a joke",
    "sit out",
    "pay",
    "practice",
    "commit",
    "politick",
    "stalk",
    "haunt",
    "check out",
    "run down",
    "lag",
    "fall behind",
    "fall back",
    "dawdle",
    "tailgate",
    "shadow",
    "ascend",
    "accommodate",
    "oblige",
    "toe the line",
    "conform to",
    "go by",
    "adhere",
    "stick",
    "keep tabs on",
    "guard",
    "proctor",
    "invigilate",
    "enter",
    "accede",
    "replace",
    "supplant",
    "supervene upon",
    "supersede",
    "vet",
    "cox",
    "follback",
  ],
  "word": "dog",
}
`;

exports[`get synonym 2`] = `
{
  "collects": [
    "true cat",
    "guy",
    "sod",
    "hombre",
    "bozo",
    "qat",
    "khat",
    "quat",
    "kat",
    "big cat",
    "cast",
    "chuck",
    "disgorge",
    "regurgitate",
    "purge",
    "honk",
    "barf",
    "regorge",
    "spew",
    "sick",
    "vomit",
    "puke",
    "be sick",
    "upchuck",
    "retch",
    "spue",
    "vomit up",
    "throw up",
  ],
  "incollects": [
    "white man",
    "babu",
    "baboo",
    "bachelor",
    "unmarried man",
    "bey",
    "boy",
    "fellow",
    "beau",
    "boyfriend",
    "swain",
    "bull",
    "strapper",
    "bruiser",
    "clotheshorse",
    "fashion plate",
    "dandy",
    "swell",
    "fop",
    "dude",
    "sheik",
    "gallant",
    "ejaculator",
    "castrate",
    "eunuch",
    "father surrogate",
    "father figure",
    "buster",
    "galoot",
    "geezer",
    "gentleman",
    "divorced man",
    "grass widower",
    "housefather",
    "hunk",
    "ex",
    "inamorato",
    "iron man",
    "ironman",
    "ironside",
    "adonis",
    "old man",
    "old boy",
    "greybeard",
    "graybeard",
    "paterfamilias",
    "patriarch",
    "ponce",
    "posseman",
    "shaver",
    "signior",
    "signor",
    "signore",
    "sir",
    "stiff",
    "stud",
    "widower",
    "widowman",
    "philanderer",
    "womanizer",
    "womaniser",
    "wonder boy",
    "golden boy",
    "young buck",
    "young man",
    "brony",
    "oppa",
    "air force officer",
    "commander",
    "gunner",
    "artilleryman",
    "cannoneer",
    "machine gunner",
    "sailor",
    "bluejacket",
    "navy man",
    "sailor boy",
    "commando",
    "ranger",
    "inductee",
    "draftee",
    "conscript",
    "enlisted person",
    "shipboard soldier",
    "devil dog",
    "leatherneck",
    "officer",
    "military officer",
    "noncombatant",
    "occupier",
    "striper",
    "vet",
    "veteran",
    "veteran soldier",
    "military volunteer",
    "voluntary",
    "volunteer",
    "humanity",
    "world",
    "mankind",
    "humankind",
    "humans",
    "human beings",
    "human race",
    "hooman",
    "black",
    "checker",
    "chequer",
    "chess piece",
    "chessman",
    "tile",
    "white",
    "crew",
    "scandalmonger",
    "telltale",
    "taleteller",
    "tattletale",
    "blabbermouth",
    "talebearer",
    "tattler",
    "yenta",
    "earful",
    "rumour",
    "rumor",
    "hearsay",
    "pipeline",
    "word of mouth",
    "grapevine",
    "malicious gossip",
    "dirt",
    "scandal",
    "talk",
    "talk of the town",
    "bruit",
    "schmoose",
    "schmooze",
    "shmooze",
    "shmoose",
    "jawbone",
    "white woman",
    "virago",
    "amazon",
    "maenad",
    "bachelor girl",
    "bachelorette",
    "baggage",
    "bar girl",
    "bas bleu",
    "bluestocking",
    "maid of honor",
    "bridesmaid",
    "broad",
    "vamp",
    "tease",
    "flirt",
    "coquette",
    "vamper",
    "prickteaser",
    "minx",
    "dame",
    "gentlewoman",
    "ma'am",
    "lady",
    "madam",
    "deb",
    "debutante",
    "grass widow",
    "divorcee",
    "dominatrix",
    "donna",
    "siren",
    "temptress",
    "femme fatale",
    "enchantress",
    "eyeful",
    "geisha girl",
    "geisha",
    "miss",
    "fille",
    "young woman",
    "girl",
    "missy",
    "young lady",
    "girlfriend",
    "lady friend",
    "gold digger",
    "gravida",
    "heroine",
    "inamorata",
    "jezebel",
    "jilt",
    "matriarch",
    "materfamilias",
    "matron",
    "mestiza",
    "kept woman",
    "mistress",
    "fancy woman",
    "mother figure",
    "nurse",
    "nanny",
    "nursemaid",
    "nullipara",
    "nymph",
    "houri",
    "nymphet",
    "old woman",
    "cocotte",
    "lady of pleasure",
    "tart",
    "prostitute",
    "woman of the street",
    "whore",
    "bawd",
    "sporting lady",
    "harlot",
    "cyprian",
    "working girl",
    "shiksa",
    "shikse",
    "dish",
    "mantrap",
    "peach",
    "beauty",
    "smasher",
    "knockout",
    "lulu",
    "sweetheart",
    "ravisher",
    "looker",
    "stunner",
    "sylph",
    "unmarried woman",
    "vestal",
    "widow woman",
    "widow",
    "wife",
    "married woman",
    "wonder woman",
    "baby mama",
    "unnie",
    "pep pill",
    "upper",
    "speed",
    "amphetamine",
    "analeptic",
    "aphrodisiac",
    "butyl nitrite",
    "isobutyl nitrite",
    "methylphenidate",
    "popper",
    "induction",
    "elicitation",
    "evocation",
    "kick",
    "turnoff",
    "negative stimulation",
    "conditioned stimulus",
    "reinforcement",
    "reinforcer",
    "reinforcing stimulus",
    "cue",
    "discriminative stimulus",
    "positive stimulus",
    "negative stimulus",
    "cowhide",
    "horsewhip",
    "knout",
    "quirt",
    "hunting crop",
    "riding crop",
    "flagellum",
    "scourge",
    "hunting whip",
    "prune whip",
    "flagellate",
    "leather",
    "switch",
    "birch",
    "urticate",
    "half track",
    "snowmobile",
    "army tank",
    "armored combat vehicle",
    "armoured combat vehicle",
    "tank",
    "stroke delimber",
    "delimber",
    "outpoint",
    "outscore",
    "walk over",
    "rack up",
    "mop up",
    "pip",
    "worst",
    "whomp",
    "get the best",
    "overcome",
    "have the best",
    "rout",
    "spreadeagle",
    "get the jump",
    "screw",
    "cheat",
    "shaft",
    "chouse",
    "jockey",
    "chicane",
    "surmount",
    "surpass",
    "exceed",
    "outstrip",
    "outdo",
    "outperform",
    "outgo",
    "outmatch",
    "get over",
    "master",
    "subdue",
    "scoop",
    "outflank",
    "best",
    "trump",
    "outfight",
    "overwhelm",
    "overmaster",
    "overpower",
    "mate",
    "checkmate",
    "immobilize",
    "immobilise",
    "outplay",
    "bat",
    "thrash",
    "lick",
    "clobber",
    "drub",
    "cream",
    "correct",
    "chasten",
    "castigate",
    "objurgate",
    "chastise",
    "tell off",
    "brush down",
    "shaving soap",
    "shaving cream",
    "bootstrap",
    "cheekpiece",
    "chin strap",
    "crupper",
    "latchet",
    "noseband",
    "nosepiece",
    "rein",
    "strop",
    "tawse",
    "hopple",
    "hobble",
    "falsie",
    "frap",
    "aerial",
    "forward pass",
    "lateral pass",
    "lateral",
    "spot pass",
    "gap",
    "col",
    "defile",
    "gorge",
    "saddle",
    "saddleback",
    "safeguard",
    "boarding card",
    "boarding pass",
    "hall pass",
    "dire straits",
    "desperate straits",
    "snap",
    "centering",
    "crack",
    "break through",
    "squeak through",
    "squeak by",
    "transit",
    "pass over",
    "cut across",
    "cross",
    "cover",
    "traverse",
    "get across",
    "cut through",
    "track",
    "infiltrate",
    "pass through",
    "run",
    "move through",
    "pass across",
    "cut",
    "crash",
    "muscle",
    "transgress",
    "overstep",
    "trespass",
    "negociate",
    "negotiate",
    "lock",
    "make",
    "work",
    "cycle",
    "fumble",
    "blunder",
    "skirt",
    "run by",
    "fly by",
    "zip by",
    "whisk by",
    "fell",
    "fly",
    "vanish",
    "give",
    "sneak",
    "slip",
    "deal",
    "deliver",
    "render",
    "turn in",
    "fork up",
    "hand over",
    "fork out",
    "fork over",
    "release",
    "relinquish",
    "free",
    "give up",
    "resign",
    "trust",
    "commit",
    "entrust",
    "intrust",
    "confide",
    "leave",
    "come",
    "radiate",
    "ray",
    "go deep",
    "go far",
    "underrun",
    "get by",
    "top",
    "clear",
    "develop",
    "break",
    "recrudesce",
    "come up",
    "arise",
    "result",
    "intervene",
    "transpire",
    "supervene",
    "go",
    "proceed",
    "fall",
    "anticipate",
    "repeat",
    "recur",
    "go off",
    "come off",
    "go over",
    "come around",
    "roll around",
    "happen",
    "materialise",
    "materialize",
    "bechance",
    "befall",
    "betide",
    "coincide",
    "concur",
    "recoil",
    "backfire",
    "backlash",
    "chance",
    "shine",
    "strike",
    "turn out",
    "synchronize",
    "contemporize",
    "synchronise",
    "contemporise",
    "soldier",
    "slum",
    "weekend",
    "holiday",
    "vacation",
    "serve",
    "do",
    "while away",
    "get through",
    "sojourn",
    "winter",
    "overwinter",
    "summer",
    "rub",
    "thread",
    "plant",
    "implant",
    "send a message",
    "relay",
    "put over",
    "call for",
    "quest",
    "ask for",
    "request",
    "bespeak",
    "telecommunicate",
    "acknowledge",
    "receipt",
    "carry",
    "return",
    "message",
    "skype",
    "nail",
    "ace",
    "sweep through",
    "pass with flying colors",
    "breeze through",
    "sail through",
    "light",
    "pass off",
    "lead",
    "approbate",
    "certificate",
    "approve",
    "sanction",
    "okay",
    "declare",
    "license",
    "certify",
    "licence",
    "commission",
    "validate",
    "formalize",
    "formalise",
    "abort",
    "stifle",
    "asphyxiate",
    "suffocate",
    "buy it",
    "pip out",
    "drown",
    "predecease",
    "famish",
    "starve",
    "yield",
    "succumb",
    "perspire",
    "sudate",
    "sweat",
    "ooze",
    "exudate",
    "ooze out",
    "exude",
    "transude",
    "puddle",
    "piddle",
    "relieve oneself",
    "piss",
    "urinate",
    "wee",
    "take a leak",
    "spend a penny",
    "pee",
    "micturate",
    "make water",
    "pass water",
    "empty",
    "evacuate",
    "void",
    "stool",
    "shit",
    "defecate",
    "crap",
    "take a shit",
    "take a crap",
    "cancel out",
    "wipe out",
    "decouple",
    "obliterate",
    "knock out",
    "cut out",
    "prune",
    "rationalize",
    "rationalise",
    "extinguish",
    "snuff out",
    "take out",
    "exclude",
    "leave off",
    "leave out",
    "omit",
    "except",
    "preclude",
    "rule out",
    "close out",
  ],
  "word": "cat",
}
`;
